api/stocks.go

```
@@ -113,11 +210,199 @@ func (s *Server) fetchStockData(ctx context.Context, symbol string) (*TimeSeries
+
+	return overview, balanceSheet, nil
+}
+
+func processBalanceSheetMetrics(
+	ctx context.Context,
+	balanceSheet *BalanceSheetMetadata,
+	metricMap map[string]model.Metric,
+	save func(model.Metric, float64),
+	symbol string,
+) {
+	if metric, ok := metricMap["Debt/Equity Ratio"]; ok {
+		value, err := calculateDebtEquityRatio(ctx, balanceSheet)
+		if err != nil {
+			log.Printf("failed to calculate Debt/Equity Ratio for stock %s: %v", symbol, err)
+			return
+		}
+		save(metric, value)
+	}
+}
+
+func processOverviewMetrics(
+	_ context.Context,
+	overview *OverviewMetadata,
+	metricMap map[string]model.Metric,
+	save func(model.Metric, float64),
+) {
+	metricExtractors := map[string]func(*OverviewMetadata) string{
+		"P/E Ratio":      func(o *OverviewMetadata) string { return o.PERatio },
+		"EPS":            func(o *OverviewMetadata) string { return o.EPS },
+		"Dividend Yield": func(o *OverviewMetadata) string { return o.DividendYield },
+		"Market Cap":     func(o *OverviewMetadata) string { return o.MarketCapitalization },
+		"Revenue Growth": func(o *OverviewMetadata) string { return o.QuarterlyRevenueGrowthYOY },
+	}
+
+	for name, metricModel := range metricMap {
+		if extractor, exists := metricExtractors[name]; exists {
+			value, err := strconv.ParseFloat(extractor(overview), 64)
+			if err != nil {
+				log.Printf("failed to convert metric %s to float: %v", name, err)
+			}
+			save(metricModel, value)
+		}
+	}
+}
+
+func (h *StockHandler) metricAdder(
+	ctx context.Context,
+	stock *model.Stock,
+	updatedMetrics *[]model.StockMetric,
+) func(metric model.Metric, value float64) {
+	return func(metric model.Metric, value float64) {
+		stockMetric := &model.StockMetric{
+			StockID:    stock.ID,
+			MetricID:   metric.ID,
+			Value:      value,
+			RecordedAt: time.Now(),
+		}
+		savedMetric, err := h.store.CreateStockMetric(ctx, stockMetric)
+		if err != nil {
+			log.Printf("failed to save metric %s for stock %s: %v", metric.Name, stock.Symbol, err)
```

Why use the `log` package? You have `https://github.com/hamba/logger` in other places, it should be given to your `Server` struct and used here instead if you can't return an error properly. But by the way, I think this should return an error when there is one 😁

---

I pretend to handle this changement in the next PR to not complexify too much this PR 👍. Therefore, I did one temporary log for this one.
For the 2nd point, if I return an error here, it will break the for loop for other metrics in `processOverviewMetrics` which are not expected to happen 🤔